File: backend/syncup/profile_links.py
"""Validation and public URL builders for profile social links."""

from __future__ import annotations

from collections.abc import Iterable
from urllib.parse import quote, urlparse

from syncup.db.models import ServiceConnection

# Manual profiles supported in Settings → Profile. Values must be a public HTTPS
# URL on the platform's own domain; this prevents profile links becoming an open
# redirect or an arbitrary-link surface.
SOCIAL_LINK_HOSTS: dict[str, frozenset[str]] = {
    "github": frozenset({"github.com", "www.github.com"}),
    "x": frozenset({"x.com", "www.x.com", "twitter.com", "www.twitter.com"}),
    "instagram": frozenset({"instagram.com", "www.instagram.com"}),
    "tiktok": frozenset({"tiktok.com", "www.tiktok.com"}),
    "youtube": frozenset({"youtube.com", "www.youtube.com", "m.youtube.com", "youtu.be"}),
    "twitch": frozenset({"twitch.tv", "www.twitch.tv"}),
    "bluesky": frozenset({"bsky.app", "www.bsky.app"}),
    "mastodon": frozenset({"mastodon.social", "www.mastodon.social"}),
    "soundcloud": frozenset({"soundcloud.com", "www.soundcloud.com"}),
}


def normalize_social_links(value: object) -> dict[str, str]:
    """Validate and normalize a map of supported profile-platform URLs."""
    if not isinstance(value, dict):
        raise ValueError("social_links must be an object")
    if len(value) > len(SOCIAL_LINK_HOSTS):
        raise ValueError("too many social links")

    normalized: dict[str, str] = {}
    for platform, raw_url in value.items():
        if platform not in SOCIAL_LINK_HOSTS:
            raise ValueError(f"unsupported social platform: {platform}")
        if not isinstance(raw_url, str):
            raise ValueError("each social link must be a URL")
        url = raw_url.strip()
        parsed = urlparse(url)
        if parsed.scheme != "https" or not parsed.hostname:
            raise ValueError("social links must use https URLs")
        if parsed.hostname.lower() not in SOCIAL_LINK_HOSTS[platform]:
            raise ValueError(f"social link must point to {platform}")
        normalized[platform] = url
    return normalized


def _service_profile_link(connection: ServiceConnection) -> tuple[str, str] | None:
    """Return a public profile link for connections with stable public URLs."""
    external_id = quote(connection.external_user_id, safe="")
    builders = {
        "anilist": lambda: ("anilist", f"https://anilist.co/user/{external_id}"),
        "lastfm": lambda: ("lastfm", f"https://www.last.fm/user/{external_id}"),
        "reddit": lambda: ("reddit", f"https://www.reddit.com/user/{external_id}"),
        "spotify": lambda: ("spotify", f"https://open.spotify.com/user/{external_id}"),
        "steam": lambda: ("steam", f"https://steamcommunity.com/profiles/{external_id}"),
        "trakt": lambda: ("trakt", f"https://trakt.tv/users/{external_id}"),
    }
    builder = builders.get(connection.service)
    return builder() if builder else None


def public_profile_links(
    social_links: dict[str, str] | None,
    connections: Iterable[ServiceConnection],
) -> dict[str, str]:
    """Merge manually-entered links with public links inferred from services."""
    result = dict(social_links or {})
    for connection in connections:
        public_link = _service_profile_link(connection)
        if public_link is not None:
            platform, url = public_link
            result[platform] = url
    return result

File: frontend/lib/constants/profile-links.ts
import {
  siBluesky,
  siGithub,
  siInstagram,
  siMastodon,
  siSoundcloud,
  siTiktok,
  siTwitch,
  siX,
  siYoutube,
} from "simple-icons";
import type { SimpleIcon } from "@/components/ui/app-icon";
import { SERVICE_BY_ID } from "@/lib/constants/services";
import type { ServiceConnection } from "@/types/api";

export interface ProfileLinkPlatform {
  id: string;
  label: string;
  brand: SimpleIcon;
}

export const MANUAL_PROFILE_LINK_PLATFORMS: ProfileLinkPlatform[] = [
  { id: "github", label: "GitHub", brand: siGithub },
  { id: "x", label: "X", brand: siX },
  { id: "instagram", label: "Instagram", brand: siInstagram },
  { id: "tiktok", label: "TikTok", brand: siTiktok },
  { id: "youtube", label: "YouTube", brand: siYoutube },
  { id: "twitch", label: "Twitch", brand: siTwitch },
  { id: "bluesky", label: "Bluesky", brand: siBluesky },
  { id: "mastodon", label: "Mastodon", brand: siMastodon },
  { id: "soundcloud", label: "SoundCloud", brand: siSoundcloud },
];

const MANUAL_PROFILE_URLS: Record<string, (username: string) => string> = {
  github: (username) => `https://github.com/${username}`,
  x: (username) => `https://x.com/${username}`,
  instagram: (username) => `https://www.instagram.com/${username}`,
  tiktok: (username) => `https://www.tiktok.com/@${username}`,
  youtube: (username) => `https://www.youtube.com/@${username}`,
  twitch: (username) => `https://www.twitch.tv/${username}`,
  bluesky: (username) => `https://bsky.app/profile/${username}`,
  mastodon: (username) => `https://mastodon.social/@${username}`,
  soundcloud: (username) => `https://soundcloud.com/${username}`,
};

/** Turn a supported platform username into its canonical public profile URL. */
export function profileUrlFromUsername(platform: string, username: string): string | null {
  const buildUrl = MANUAL_PROFILE_URLS[platform];
  if (!buildUrl) return null;
  const normalizedUsername = username.trim().replace(/^@/, "");
  return normalizedUsername ? buildUrl(encodeURIComponent(normalizedUsername)) : null;
}

/** Extract the visible account identifier from a canonical public profile URL. */
export function profileUsernameFromUrl(url: string): string | null {
  try {
    const segments = new URL(url).pathname.split("/").filter(Boolean);
    const username = segments.at(-1);
    return username ? decodeURIComponent(username) : null;
  } catch {
    return null;
  }
}

const SERVICE_PROFILE_URLS: Record<string, (externalId: string) => string> = {
  anilist: (id) => `https://anilist.co/user/${encodeURIComponent(id)}`,
  lastfm: (id) => `https://www.last.fm/user/${encodeURIComponent(id)}`,
  reddit: (id) => `https://www.reddit.com/user/${encodeURIComponent(id)}`,
  spotify: (id) => `https://open.spotify.com/user/${encodeURIComponent(id)}`,
  steam: (id) => `https://steamcommunity.com/profiles/${encodeURIComponent(id)}`,
  trakt: (id) => `https://trakt.tv/users/${encodeURIComponent(id)}`,
};

export function profileLinkPlatform(id: string): ProfileLinkPlatform | undefined {
  const manual = MANUAL_PROFILE_LINK_PLATFORMS.find((platform) => platform.id === id);
  if (manual) return manual;
  const service = SERVICE_BY_ID.get(id);
  return service
    ? { id: service.id, label: service.name, brand: service.brand }
    : undefined;
}

/** Add links inferred from services that expose a stable public profile URL. */
export function buildOwnProfileLinks(
  socialLinks: Record<string, string> | null,
  connections: ServiceConnection[],
) {
  const links = { ...(socialLinks ?? {}) };
  for (const connection of connections) {
    const buildUrl = SERVICE_PROFILE_URLS[connection.service];
    if (buildUrl) links[connection.service] = buildUrl(connection.external_user_id);
  }
  return links;
}
